hawkbit-autoconfigure/src/main/java/org/eclipse/hawkbit/autoconfigure/security/InMemoryUserManagementAutoConfiguration.java

```diff
@@ -11,129 +11,155 @@
 public class InMemoryUserManagementAutoConfiguration extends GlobalAuthenticationConfigurerAdapter {
 
     private static final String DEFAULT_TENANT = "DEFAULT";
 
-    private final SecurityProperties securityProperties;
-
-    private final MultiUserProperties multiUserProperties;
+    private final UserDetailsService userDetailsService;
 
-    InMemoryUserManagementAutoConfiguration(final SecurityProperties securityProperties,
-            final MultiUserProperties multiUserProperties) {
-        this.securityProperties = securityProperties;
-        this.multiUserProperties = multiUserProperties;
+    InMemoryUserManagementAutoConfiguration(
+            final SecurityProperties securityProperties,
+            final TenantAwareUserProperties tenantAwareUserProperties,
+            final Optional<PasswordEncoder> passwordEncoder) {
+        userDetailsService = userDetailsService(
+                securityProperties, tenantAwareUserProperties, passwordEncoder.orElse(null));
     }
 
     @Override
-    public void configure(final AuthenticationManagerBuilder auth) throws Exception {
+    public void configure(final AuthenticationManagerBuilder auth) {
         final DaoAuthenticationProvider userDaoAuthenticationProvider = new TenantDaoAuthenticationProvider();
-        userDaoAuthenticationProvider.setUserDetailsService(userDetailsService());
+        userDaoAuthenticationProvider.setUserDetailsService(userDetailsService);
         auth.authenticationProvider(userDaoAuthenticationProvider);
     }
 
-    /**
-     * @return the user details service to load a user from memory user manager.
-     */
-    @Bean
-    @ConditionalOnMissingBean
-    UserDetailsService userDetailsService() {
-
-        final List<UserPrincipal> userPrincipals = new ArrayList<>();
-        for (final MultiUserProperties.User user : multiUserProperties.getUsers()) {
-            final List<String> permissions = user.getPermissions();
-            List<GrantedAuthority> authorityList;
-            // Allows ALL as a shorthand for all permissions
-            if (permissions.size() == 1 && "ALL".equals(permissions.get(0))) {
-                authorityList = PermissionUtils.createAllAuthorityList();
-            } else {
-                authorityList = createAuthoritiesFromList(permissions);
-            }
-
-            final UserPrincipal userPrincipal = new UserPrincipal(user.getUsername(), user.getPassword(),
-                    user.getFirstname(), user.getLastname(), user.getUsername(), user.getEmail(), DEFAULT_TENANT,
-                    authorityList);
+    private static UserDetailsService userDetailsService(
+            final SecurityProperties securityProperties,
+            final TenantAwareUserProperties tenantAwareUserProperties,
+            final PasswordEncoder passwordEncoder) {
+        final List<User> userPrincipals = new ArrayList<>();
+        tenantAwareUserProperties.getUsers().forEach((username, user) -> {
```

Comment:
here tenantAwareUserProperties#getUsers is supposed to be a map<String, Users[]> where key is tenant and values are colleciton of users, right :?
but here key of the map is taken as username, rather than tenant

Reply:
have we tested this with different usernames ? maybe some symbols are not allowed here, or are not properly parsed from spring properties ?

Reply:
since it is default "spring" approach (e.g. spring.security.user.<username>) all usernames that are accepted as spring users will be accepted here - which shall be enough